base_layer/wallet/tests/support/rpc.rs

```diff
@@ -0,0 +1,266 @@
+// Copyright 2020. The Tari Project
+//
+// Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
+// following conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following
+// disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+// following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+// products derived from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use std::{
+    convert::TryFrom,
+    sync::{Arc, Mutex},
+    time::{Duration, Instant},
+};
+use tari_comms::protocol::rpc::{Request, Response, RpcStatus};
+use tari_core::{
+    base_node::{
+        proto::wallet_response::{TxLocation, TxQueryResponse, TxSubmissionRejectionReason, TxSubmissionResponse},
+        rpc::BaseNodeWalletService,
+    },
+    proto::generated::{
+        base_node::{TxQueryResponse as TxQueryResponseProto, TxSubmissionResponse as TxSubmissionResponseProto},
+        types::{Signature as SignatureProto, Transaction as TransactionProto},
+    },
+    transactions::{transaction::Transaction, types::Signature},
+};
+use tokio::time::delay_for;
+
+/// This macro unlocks a Mutex or RwLock. If the lock is
+/// poisoned (i.e. panic while unlocked) the last value
+/// before the panic is used.
+macro_rules! acquire_lock {
+    ($e:expr, $m:ident) => {
+        match $e.$m() {
+            Ok(lock) => lock,
+            Err(poisoned) => {
+                log::warn!(target: "wallet", "Lock has been POISONED and will be silently recovered");
+                poisoned.into_inner()
+            },
+        }
+    };
+    ($e:expr) => {
+        acquire_lock!($e, lock)
+    };
+}
+
+#[derive(Clone)]
+pub struct BaseNodeWalletRpcMockState {
```

Comment:
👀 
@delta1 this module might be of interest on the work you are currently doing as it shows how to set up a mock RPC Server to test you RPC client calls against. Tests at the bottom show how it is used.

Reply:
Brilliant thanks @philipr-za will take a look and also review ASAP